core/state/pruner/pruner.go

```diff
@@ -427,23 +427,46 @@ func dumpRawTrieDescendants(db ethdb.Database, root common.Hash, output *stateBl
 						return
 					}
 					var processedNodes uint64
-					for storageIt.Next(true) {
-						storageTrieHash := storageIt.Hash()
-						if storageTrieHash != (common.Hash{}) {
-							// The inner bloomfilter library has a mutex so concurrency is fine here
-							err = output.Put(storageTrieHash.Bytes(), nil)
+					isNext := true
+					storageTrieHashes := make(chan common.Hash, threads)
+					go func() {
+						for isNext {
+							err = <-results
 							if err != nil {
 								return
 							}
+							go func() {
+								threadsRunning.Add(1)
+								defer threadsRunning.Add(-1)
+								var err error
+								defer func() {
+									results <- err
+								}()
+								for storageTrieHash := range storageTrieHashes {
+									if storageTrieHash != (common.Hash{}) {
+										// The inner bloomfilter library has a mutex so concurrency is fine here
+										err = output.Put(storageTrieHash.Bytes(), nil)
+										if err != nil {
+											return
+										}
+									}
+									processedNodes++
+									if time.Since(threadLastLog) > 5*time.Minute {
+										elapsedTotal := time.Since(startedAt)
+										elapsedThread := time.Since(threadStartedAt)
+										log.Info("traversing trie database - traversing storage trie taking long", "key", key, "elapsedTotal", elapsedTotal, "elapsedThread", elapsedThread, "processedNodes", processedNodes, "threadsRunning", threadsRunning.Load())
+										threadLastLog = time.Now()
+									}
+								}
+							}()
 						}
-						processedNodes++
-						if time.Since(threadLastLog) > 5*time.Minute {
-							elapsedTotal := time.Since(startedAt)
-							elapsedThread := time.Since(threadStartedAt)
-							log.Info("traversing trie database - traversing storage trie taking long", "key", key, "elapsedTotal", elapsedTotal, "elapsedThread", elapsedThread, "processedNodes", processedNodes, "threadsRunning", threadsRunning.Load())
-							threadLastLog = time.Now()
-						}
+					}()
+					for isNext {
+						isNext = storageIt.Next(true)
```

Comment:
I don't think this helps. You still have only one thread traversing all storage, which Is the bottleneck.
Creating parallelism should be done using the parameter in this line:
```
storageTr.NodeIterator(nil)
```
The nil is "from which key to start". Instead of that, have multiple threads start from multiple values and have each stop where the next thread starts. I assume (not sure) that it's enough to give the first part of the key (you obviously don't know the full key at this time)